languages/csharp/udSDK.cs

```diff
@@ -173,6 +176,62 @@ public void Connect(string pURL, string pApplicationName, string pKey)
         throw new Exception("Unknown error occurred, please try again later.");
     }
 
+    public void ConnectStart(string pURL, string pApplicationName, string pApplicationVersion, ref string ppApprovePath, ref string ppApproveCode)
+    {
+      udError error = udError.udE_Failure;
+
+      error = udContext_TryResume(ref pContext, pURL, pApplicationName, null, 0); //Set to 1 to try use the dongle
+
+      if (error != udError.udE_Success)
+      {
+        IntPtr approvePath = IntPtr.Zero;
+        IntPtr approveCode = IntPtr.Zero;
+        error = udContext_ConnectStart(ref pContextPartial, pURL, pApplicationName, pApplicationVersion, ref approvePath, ref approveCode);
+        ppApprovePath = Marshal.PtrToStringUTF8(approvePath);
+        ppApproveCode = Marshal.PtrToStringUTF8(approveCode);
+      }
+
+      if (error == udError.udE_ConnectionFailure)
+        throw new Exception("Could not connect to server.");
+      else if (error == udError.udE_AuthFailure)
+        throw new Exception("Username or Password incorrect.");
+      else if (error == udError.udE_OutOfSync)
+        throw new Exception("Your clock doesn't match the remote server clock.");
+      else if (error == udError.udE_SecurityFailure)
+        throw new Exception("Could not open a secure channel to the server.");
+      else if (error == udError.udE_ServerFailure)
+        throw new Exception("Unable to negotiate with server, please confirm the server address");
+      else if (error != udError.udE_Success)
+        throw new Exception("Unknown error occurred, please try again later.");
+    }
+
+    public void ConnectComplete()
+    {
+      udError error = udError.udE_Failure;
+
+      error = udContext_ConnectComplete(ref pContext, ref pContextPartial);
+
```

Comment:
This method of converting return codes to exceptions seems brittle- are we better off defining a separate Exception class that converts the return code to an exception?

Reply:
something like 
```
if(returnCode !=udE_Success)
{
 throw new udException(returnCode);
}
```
Would be cleaner